python-connectors/kayrros_retrieve-dataset/connector.py

```diff
@@ -0,0 +1,122 @@
+# This file is the actual code for the custom Python dataset kayrros_retrieve-dataset
+
+from dataiku.connector import Connector
+from utils.authentification import get_headers
+import pandas as pd
+import requests
+import logging
+
+logger = logging.getLogger(__name__) 
+
+
+class MyConnector(Connector):
+
+    # A custom Python dataset is a subclass of connector.
+
+    def __init__(self, config, plugin_config):
+
+        Connector.__init__(self, config, plugin_config) 
+
+        # Retrieve credentials and token
+
+        self.username = config.get("username", "")
+        if not self.username:
+            raise ValueError("A Kayrros account is necessary to fetch the data. Please provide a username.")
+
+        self.password = config.get("password", "")
+        if not self.password:
+            raise ValueError("A Kayrros account is necessary to fetch the data. Please provide a password.")
+
+        self.headers = get_headers(self.username, self.password)
+
+        # Retrieve ids of collection and dataset
+
+        self.collection_id = config.get("collection_id", "")
+        if not self.collection_id:
+            raise ValueError("Choosing a collection is necessary to fetch the data. Please provide one in the collection field.")
+
+        self.dataset_id = config.get("dataset_id", "")
+        if not self.password:
+            raise ValueError("Choosing a dataset is necessary to fetch the data. Please provide one in the collection field.")
+
+    def get_read_schema(self):
+
+        # We don't specify a schema here, so DSS will infer the schema
+        # from the columns actually returned by the generate_rows method
+
+        return None
+
+    def generate_rows(self, dataset_schema=None, dataset_partitioning=None, partition_id=None, records_limit=-1):
+
+        url_asset = "https://platform.api.kayrros.com/v1/processingresult/dataset/" + self.dataset_id
+
+        try:        
+            response = requests.post(url_asset, headers=self.headers)
+            response.raise_for_status()
+
+        except requests.exceptions.RequestException as error:
+            logger.exception("Dataset could not be retrieved because of the following error:\n {}".format(error))
+
```

Comment:
You need to raise an exception here, otherwise the error message visible to the user will be cryptic (something like `Test failed: Failed to read from python dataset : <class 'json.decoder.JSONDecodeError'> : Expecting value: line 1 column 1 (char 0) `), while the actual reason for the crash (here an HTTP 504 timeout) is buried in the logs...

Reply:
solved: [commit](https://github.com/dataiku/dss-plugin-kayrros/commit/aaa0143f806dba4d8c7bb2e93fc123f0efa139ff)